### PR TITLE
better edges when arrow_shift = :end (fixes #171)

### LINKED ISSUE #171: arrow_shift = :end works poorly with larger edge width
Here's an example with edge width 4.

0125-11073728-problem

Automatically handling this is probably too much to ask. But perhaps we could have an argument that allowed us to fine-tune the positions starting with `arrow_shift = :end`? I think the correct way to handle this would be to simply move the end point of the arrow by some fixed pixel value, (I don't think it should scale with the radius of the destination node).

### PR DESCRIPTION
**Pull request by fredcallaway**
One test fails, but it seems to also fail without my modification (`no method matching discretize(::BezierPath)`)

This is only tested for 2d and circular nodes.

image

### CODE REVIEW

File: src/beziercurves.jl [hunk 1]
```diff
@@ -43,6 +43,14 @@ ptype(::Union{AbstractPath{PT}, Type{<:AbstractPath{PT}}}) where {PT} = PT
 straighten(l::Line) = l
 straighten(p::BezierPath) = Line(interpolate(p,0.0), interpolate(p,1.0))
 
+adjust_endpoint(line::Line, p) = Line(line.p0, p)
+function adjust_endpoint(path::BezierPath, p)
+    commands = copy(path.commands)
+    c = commands[end]
+    commands[end] = CurveTo(c.c1, c.c2, p)
```
Comment (hexaeder) [inline src/beziercurves.jl:50]: this should be something like
```suggestion
    commands[end] = adjust_endpoint(commands[end], p)
```
with additional overloads
```julia
adjust_endpoint(::MoveTo, p) = MoveTo(p)
adjust_endpoint(::LineTo, p) = LineTo(p)
adjust_endpoint(c::CurveTo, p) = CurveTo(c.c1, c.c2, p)
```
such that it works on other curve types as well.

**Comment by hexaeder** [comment]
Thanks for the PR! Nice and simple solution!

You changed the order of the plots in the recipe, that breaks the `get_edge_plot` "logic" (which is stupid design on my part anyway). Maybe you can change the definition of the `get_edge_plot` function such that it looks for the first subplot with `plot isa EdgePlot`  instead of choosing by index?

File: src/recipes.jl [hunk 2]
```diff
@@ -325,6 +319,22 @@ function Makie.plot!(gp::GraphPlot)
         visible = arrow_show_m,
         gp.arrow_attr...)
 
+
+    # update edge paths to line up with arrow heads if arrow_shift = :end
+    edge_paths = gp[:edge_paths] = @lift map($init_edge_paths, $arrow_pos, eachindex($arrow_pos)) do ep, ap, i
```
Comment (hexaeder) [inline src/recipes.jl:324]: i think `@lift map` is redundent as observables.jl overloads map directly https://juliagizmos.github.io/Observables.jl/stable/#Base.map-Union{Tuple{F},%20Tuple{F,%20AbstractObservable,%20Vararg{Any}}}%20where%20F

File: src/recipes.jl [hunk 3]
```diff
@@ -753,6 +763,36 @@ lands on the surface of the destination node.
...
+                t = 0.5
+            end
+        end
+        arrow_shift[i] = t
+    end
+
+    return arrow_shift
+end
+
+function update_paths_ends(g, gp, edge_paths::Vector{<:AbstractPath{PT}}, arrow_pos) where {PT}
```
Comment (hexaeder) [inline src/recipes.jl:793]: seems like leftover code?

**Comment by fredcallaway** [comment]
I think I resolved all the points above. I also figured out how to actually run the tests (oops). I'm getting an error in referencetests.jl though and I'm having a hard time tracking it down because of the way files are being generated (I can't figure out where the error is actually occurring (the stack trace is pointing to an empty line, could the reftests.jl I see in my directory might be different than the one that was actually run???)

**Comment by hexaeder** [comment]
So what happes in the test is that all the examples from `docs/examples` are executed and the plots within the examples will be saved as reference images. Apparently, there is a problem with on of them in the `reftests.jl` example file.

I wasn't able to debug it yet, but the problem is this part of that file:

https://github.com/MakieOrg/GraphMakie.jl/blob/543cc34f391e7de40658ed6c367037ca4e087392/docs/examples/reftests.jl#L112-L130

were we use symbols to define various linestyles. I have no idea why that does not work on your PR but on master though, needs debugging. Preferably by manual evaluating that example file.

**Comment by hexaeder** [comment]
Found it, you didn't introduce any bug but instead found one. Should be fixed now. I also restored the ordering of of edge and arrow plot, such that the arrows will overdraw the edges.

However I noticed that changing the endpoint of a bezier curve also changes the curve in general. Which means the arrows will not be perfectly tangential anymore. In reference picture 18 one can see it in the selfloop of the uppermost node:

reftests jl-18

reftests jl-18+

But do we care?

**Pull request merged**